module/zcommon/zfs_uio.c

```diff
@@ -79,8 +80,24 @@ uiomove_iov(void *p, size_t n, enum uio_rw rw, struct uio *uio)
 				if (copy_to_user(iov->iov_base+skip, p, cnt))
 					return (EFAULT);
 			} else {
-				if (copy_from_user(p, iov->iov_base+skip, cnt))
-					return (EFAULT);
+				if (uio->uio_fault_disable) {
+					if (!access_ok(VERIFY_READ,
+					    (iov->iov_base + skip), cnt)) {
+						return (EFAULT);
+					}
+
+					pagefault_disable();
+					if (__copy_from_user_inatomic(p,
+					    (iov->iov_base + skip), cnt)) {
+						pagefault_enable();
+						return (EFAULT);
```

Comment:
The problem with returning EFAULT here is that the caller can't differentiate if the error is caused by access_ok or __copy_from_user_inatomic. And even if the error is caused by __copy_from_user_inatomic, you still need to figure out if the faulted address is a valid address. Otherwise, you'll just be in an infinite loop.

I see the current uio_prefaultpages doesn't return anything. Perhaps we should make it return EFAULT, so we would know if the page fault can be resolved or not.

Reply:
Thanks for your comments. 

The fix actually takes effect ONLY in zfs_write context during calling uiomove_iov (there is a flag uio_pagefault_disable to control that). So the pages should have already been faulted at least once previously . If the referenced address is invalid, panic would occur in uio_prefaultpages before entering uiomove(). Frankly, access_ok() here in uiomove_iov() is not mandatory in my mind.

From the code perspective,  both access_ok() and __copy_from_user_inatomic() could return EFAULT. But actually, access_ok() would never return EFAULT in this case.

If you prefer more sophisticate checking and my understanding is correct, you suggest make following modifications:
1. access_ok() check in uio_prefaultpages.
2. we remove access_ok() from uiomove_iov().

Is my understanding correct?

Reply:
Hi @javenwu 
Sorry my previous comment was not very clear.

The address in uio is from user input, so we can't assume validity about it. There are roughly three kinds of address we can get.
1. An address that is in the userspace range (e.g. <= 0x7FFFFFFFFFFF for amd64), and is properly mapped. (For the sake of discussion we'll also assume it's not on a physical page yet)
2. An address that is in the userspace range, but is not mapped to anything.
3. An address that is not in the userspace range.

In these 3 cases, only the first only can return success. For 2 and 3, we must return EFAULT to userspace.

To catch 3, we need `access_ok()`, access_ok is implied in more common kernel APIs handling user address, like `copy_from_user`, but in this case you are using `__copy_from_user_inatomic()` so we need to check it. And you must check it before calling `__copy_from_user_inatomic()`, otherwise the user can pass in a kernel address, and you'll have a info leak. In the current patch version, you have `access_ok()` in correct position. The only thing missing is you need to return EFAULT to userspace in this case and not continue looping.

As for 1 and 2, both will pass `access_ok()`, and both are safe to be called with `__copy_from_user_inatomic()` as both will get EFAULT. However, the difference is that after `uio_prefaultpages()`, 1 will successfully fault in a page where 2 will never. So if we success on `access_ok()` and get EFAULT on `__copy_from_user_inatomic()`, we still don't really know the address is case 1 or case 2.

The current implementation of `uio_prefaultpages()` doesn't return any thing, but internally it will notice when a page fault failed and just return immediately. So what we need to do here is to return this info when a page fault failed. This way we can differentiate if the address is case 1 or 2. And if it's case 2, we need to return EFAULT to userspace and not continue looping.

Reply:
Thanks for the clarification. You are right, I neglect the case 2. We should check the case 2 and break  the loop.